Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm

```diff
@@ -4960,8 +4960,18 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
     context.contextAfter = makeString(rangeOfInterestInSelection.end, contextAfterEnd);
     if (auto compositionVisiblePositionRange = makeVisiblePositionRange(compositionRange); intersects(rangeOfInterest, compositionVisiblePositionRange)) {
         context.markedText = makeString(compositionVisiblePositionRange.start, compositionVisiblePositionRange.end);
-        context.selectedRangeInMarkedText.location = distanceBetweenPositions(rangeOfInterestInSelection.start, compositionVisiblePositionRange.start);
-        context.selectedRangeInMarkedText.length = [context.selectedText.string length];
+        auto markedTextLength = context.markedText.string.length();
+
+        ptrdiff_t distanceToSelectionStart = distanceBetweenPositions(rangeOfInterestInSelection.start, compositionVisiblePositionRange.start);
+        ptrdiff_t distanceToSelectionEnd = distanceToSelectionStart + [context.selectedText.string length];
+
+        distanceToSelectionStart = clampTo<ptrdiff_t>(distanceToSelectionStart, 0, markedTextLength);
+        distanceToSelectionEnd = clampTo<ptrdiff_t>(distanceToSelectionEnd, 0, markedTextLength);
```

Comment:
Should we make sure start < end as well?

Reply:
Oh, I think that because `[context.selectedText.string length] >= 0`, it's guaranteed that `distanceToSelectionStart  <= distanceToSelectionEnd`. I can add a `RELEASE_ASSERT(distanceToSelectionStart  <= distanceToSelectionEnd);` here to affirm (and clarify) this.

Thank you for the reviews!